pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/merge_rollup/MergeRollupTaskGenerator.java

```diff
@@ -0,0 +1,454 @@
+              TimeUtils.convertPeriodToMillis(e2.getValue().get(MinionConstants.MergeTask.BUCKET_TIME_PERIOD_KEY))))
+          .collect(Collectors.toList());
+
+      // Schedule tasks from lowest to highest merge level
+      String lowerMergeLevel = null;
+      long lowerMergeLevelWatermarkMs = -1;
+
+      for (Map.Entry<String, Map<String, String>> entry : sortedMergeLevelConfigs) {
+        String mergeLevel = entry.getKey();
+        Map<String, String> mergeConfig = entry.getValue();
+
+        // Get the bucket size and buffer
+        long bucketMs =
+            TimeUtils.convertPeriodToMillis(mergeConfig.get(MinionConstants.MergeTask.BUCKET_TIME_PERIOD_KEY));
+        long bufferMs =
+            TimeUtils.convertPeriodToMillis(mergeConfig.get(MinionConstants.MergeTask.BUFFER_TIME_PERIOD_KEY));
+
+        // Get watermark from MergeRollupTaskMetadata ZNode.
+        // windowStartMs = watermarkMs, windowEndMs = windowStartMs + bucketTimeMs
+        long waterMarkMs =
+            getWatermarkMs(offlineTableName, preSelectedSegmentsForOfflineTable.first().getStartTimeMs(), bucketMs,
+                mergeLevel);
+        long windowStartMs = waterMarkMs;
+        long windowEndMs = windowStartMs + bucketMs;
+
+        // Skip scheduling if there's incomplete task for current mergeLevel
+        if (inCompleteMergeLevelToTaskConfig.containsKey(mergeLevel)) {
+          LOGGER.info(
+              "Found incomplete tasks: {} of merge level: {} for the same table: {}, Skipping task generation: {}",
+              inCompleteMergeLevelToTaskConfig.get(mergeLevel), mergeLevel, offlineTableName, taskType);
+          lowerMergeLevel = mergeLevel;
+          lowerMergeLevelWatermarkMs = waterMarkMs;
+          continue;
+        }
+
+        if (!isValidMergeWindowEndTime(windowEndMs, bufferMs, lowerMergeLevel, lowerMergeLevelWatermarkMs)) {
+          LOGGER.info(
+              "Window with start: {} and end: {} of mergeLevel: {} is not a valid merge window, Skipping task generation: {}",
+              windowStartMs, windowEndMs, mergeLevel, taskType);
+          lowerMergeLevel = mergeLevel;
+          lowerMergeLevelWatermarkMs = waterMarkMs;
+          continue;
+        }
+
+        // Find all segments overlapping with the merge window, if all overlapping segments are merged, bump up the target window
+        List<OfflineSegmentZKMetadata> selectedSegments = new ArrayList<>();
+        boolean hasUnMergedSegments = false;
+        boolean isValidMergeWindow = true;
+
+        for (OfflineSegmentZKMetadata preSelectedSegment : preSelectedSegmentsForOfflineTable) {
+          if (windowStartMs <= preSelectedSegment.getEndTimeMs() && preSelectedSegment.getStartTimeMs() < windowEndMs) {
+            // For segments overlapping with current window, add to the result list
+            selectedSegments.add(preSelectedSegment);
```

Comment:
Do we need an extra filter that filters segments from the previous merge level? Eventually, we need to make sure the following happens:

Let's say we have merge level with `1d, 30d`.

1. original segments -> merge to `1d`
2. merged segments from step 1 (`1d`)  -> merge to `30d`

Reply:
We don't need another filter here, we won't have any pre-merged segments from the previous merge level that's overlapping with the current merge window. Say we are merging 1d segments to 30d, the merge window end time will always be earlier than 1d level watermark. There won't be any original segments before the 1d watermark since they are merged already.